src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +141,206 @@ public interface IFileWatchingToken
         {
         }
 
+        private readonly struct WatcherOperation
+        {
+            private readonly Kind _kind;
+
+            private readonly string _directory;
+            private readonly string? _filter;
+            private readonly _VSFILECHANGEFLAGS _fileChangeFlags;
+            private readonly IVsFreeThreadedFileChangeEvents2 _sink;
+            private readonly List<uint> _cookies;
+
+            private readonly IFileWatchingToken _token;
+            private readonly IEnumerable<IFileWatchingToken> _tokens;
+
+            private WatcherOperation(Kind kind)
+            {
+                Contract.ThrowIfFalse(kind is Kind.None);
+                _kind = kind;
+
+                // Other watching fields are not used for this kind
+                _directory = null!;
```

Comment:
instead of making the fields non-nullable, make them nullable and have an IsXXX check with MemberNotNull attributes and have the appropraiate methods make teh appropriate contract calls which can then inform the flow checker which fields will then really be non-null or not.

Reply:
I opted against this because it doesn't work well with switch expressions.

Reply:
true.  that's a pity.  @jcouv i'd be interested in working on a proposal with you where NRT ties the state of value to an enum.  This would match some features that TypeScript has, and woudl work really nice for this sort of 'flattened struct' pattern.